NPE：ERROR 2013 (HY000): Lost connection to MySQL server during query

## Bug Report

**For English only**, other languages will not accept.

Before report a bug, make sure you have:

- Searched open and closed [GitHub issues](https://github.com/apache/shardingsphere/issues).
- Read documentation: [ShardingSphere Doc](https://shardingsphere.apache.org/document/current/en/overview).

Please pay attention on issues you submitted, because we maybe need more details. 
If no response anymore and we cannot reproduce it on current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Which version of ShardingSphere did you use?
master
### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
 ShardingSphere-Proxy
### Expected behavior
```
mysql> show sharding table rules;

sharding rule ~
```
### Actual behavior
```
mysql> show sharding table rules;
ERROR 2013 (HY000): Lost connection to MySQL server during query
```
### Reason analyze (If you can)
Maybe DistSQL have BUG？
keyGenerator in yaml don't hava verify?
### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
##### use yaml
```
databaseName: sharding_db

dataSources:
  ds_0:
    url: jdbc:mysql://127.0.0.1:3306/ds_0?serverTimezone=UTC&useSSL=false&allowPublicKeyRetrieval=true
    username: root
    password: root
    maxPoolSize: 40
    minPoolSize: 1
  ds_1:
    url: jdbc:mysql://127.0.0.1:3306/ds_1?serverTimezone=UTC&useSSL=false&allowPublicKeyRetrieval=true
    username: root
    password: root
    maxPoolSize: 40
    minPoolSize: 1
  ds_2:
    url: jdbc:mysql://127.0.0.1:3306/ds_2?serverTimezone=UTC&useSSL=false&allowPublicKeyRetrieval=true
    username: root
    password: root
    maxPoolSize: 40
    minPoolSize: 1
  ds_3:
    url: jdbc:mysql://127.0.0.1:3306/ds_3?serverTimezone=UTC&useSSL=false&allowPublicKeyRetrieval=true
    username: root
    password: root
    maxPoolSize: 5
    minPoolSize: 1
  ds_4:
    url: jdbc:mysql://127.0.0.1:3306/ds_4?serverTimezone=UTC&useSSL=false&allowPublicKeyRetrieval=true
    username: root
    password: root
    maxPoolSize: 5
    minPoolSize: 1

rules:
  - !SHARDING
    autoTables:
      t_order:
        actualDataSources: source_readwrite
        keyGenerateStrategy:
          column: order_id
          keyGeneratorName: t_order_snowflake
        logicTable: t_order
        shardingStrategy:
          standard:
            shardingAlgorithmName: t_order_hash_mod
            shardingColumn: order_id
    shardingAlgorithms:
      t_order_inline:
        props:
          algorithm-expression: t_order_${order_id % 2}
        type: INLINE
      t_order_hash_mod:
        props:
          sharding-count: '6'
        type: hash_mod
    keyGenerators:
      snowflake:
        type: SNOWFLAKE

    scalingName: default_scaling
    scaling:
      default_scaling:
        input:
          workerThread: 40
          batchSize: 1000
        output:
          workerThread: 40
          batchSize: 1000
        streamChannel:
          type: MEMORY
          props:
            block-queue-size: 10000
        dataConsistencyChecker:
          type: DATA_MATCH
          props:
            chunk-size: 1000

  - !READWRITE_SPLITTING
    dataSources:
      source_readwrite:
        type: Static
        props:
          write-data-source-name: ds_0
          read-data-source-names: ds_1
        loadBalancerName: random
      target_readwrite:
        type: Static
        props:
          write-data-source-name: ds_2
          read-data-source-names: ds_3
        loadBalancerName: random
    loadBalancers:
      random:
        type: RANDOM
```
### Example codes for reproduce this issue (such as a github link).

@strongduanmu  @RaigorJiang  @azexcy 
plz have a look
I use Incorrect configuration triggered the problem.

```
[ERROR] 2022-07-01 17:21:06.733 [ShardingSphere-Command-4] o.a.s.p.f.c.CommandExecutorTask - Exception occur:
java.lang.NullPointerException: null
        at org.apache.shardingsphere.sharding.distsql.handler.query.ShardingTableRuleQueryResultSet.getKeyGeneratorType(ShardingTableRuleQueryResultSet.java:211)
        at org.apache.shardingsphere.sharding.distsql.handler.query.ShardingTableRuleQueryResultSet.buildAutoTableRowData(ShardingTableRuleQueryResultSet.java:137)
        at org.apache.shardingsphere.sharding.distsql.handler.query.ShardingTableRuleQueryResultSet.getRowData(ShardingTableRuleQueryResultSet.java:95)
        at org.apache.shardingsphere.proxy.backend.text.distsql.rql.RQLBackendHandler.getRowData(RQLBackendHandler.java:61)
        at org.apache.shardingsphere.proxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.getQueryRowPacket(MySQLComQueryPacketExecutor.java:131)
        at org.apache.shardingsphere.proxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.getQueryRowPacket(MySQLComQueryPacketExecutor.java:59)
        at org.apache.shardingsphere.proxy.frontend.mysql.command.MySQLCommandExecuteEngine.writeQueryData(MySQLCommandExecuteEngine.java:90)
        at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:113)
        at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:77)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
[ERROR] 2022-07-01 17:21:37.785 [ShardingSphere-Command-4] o.a.s.p.f.c.CommandExecutorTask - Exception occur:
java.lang.NullPointerException: null
        at org.apache.shardingsphere.sharding.distsql.handler.query.ShardingTableRuleQueryResultSet.getKeyGeneratorType(ShardingTableRuleQueryResultSet.java:211)
        at org.apache.shardingsphere.sharding.distsql.handler.query.ShardingTableRuleQueryResultSet.buildAutoTableRowData(ShardingTableRuleQueryResultSet.java:137)
        at org.apache.shardingsphere.sharding.distsql.handler.query.ShardingTableRuleQueryResultSet.getRowData(ShardingTableRuleQueryResultSet.java:95)
        at org.apache.shardingsphere.proxy.backend.text.distsql.rql.RQLBackendHandler.getRowData(RQLBackendHandler.java:61)
        at org.apache.shardingsphere.proxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.getQueryRowPacket(MySQLComQueryPacketExecutor.java:131)
        at org.apache.shardingsphere.proxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.getQueryRowPacket(MySQLComQueryPacketExecutor.java:59)
        at org.apache.shardingsphere.proxy.frontend.mysql.command.MySQLCommandExecuteEngine.writeQueryData(MySQLCommandExecuteEngine.java:90)
        at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:113)
        at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:77)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

--- a/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/builder/AlgorithmProvidedShardingRuleBuilder.java
+++ b/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/builder/AlgorithmProvidedShardingRuleBuilder.java
@@ -22,8 +22,14 @@ import org.apache.shardingsphere.infra.instance.InstanceContext;
 import org.apache.shardingsphere.infra.rule.ShardingSphereRule;
 import org.apache.shardingsphere.infra.rule.builder.database.DatabaseRuleBuilder;
 import org.apache.shardingsphere.sharding.algorithm.config.AlgorithmProvidedShardingRuleConfiguration;
+import org.apache.shardingsphere.sharding.api.config.rule.ShardingAutoTableRuleConfiguration;
+import org.apache.shardingsphere.sharding.api.config.rule.ShardingTableRuleConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.keygen.KeyGenerateStrategyConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.sharding.ShardingStrategyConfiguration;
 import org.apache.shardingsphere.sharding.constant.ShardingOrder;
 import org.apache.shardingsphere.sharding.rule.ShardingRule;
+import org.apache.shardingsphere.sharding.spi.KeyGenerateAlgorithm;
+import org.apache.shardingsphere.sharding.spi.ShardingAlgorithm;
 
 import javax.sql.DataSource;
 import java.util.Collection;
@@ -38,7 +44,44 @@ public final class AlgorithmProvidedShardingRuleBuilder implements DatabaseRuleB
     public ShardingRule build(final AlgorithmProvidedShardingRuleConfiguration config, final String databaseName,
                               final Map<String, DataSource> dataSources, final Collection<ShardingSphereRule> builtRules, final InstanceContext instanceContext) {
         Preconditions.checkArgument(null != dataSources && !dataSources.isEmpty(), "Data sources cannot be empty.");
+        Preconditions.checkArgument(isValidRuleConfiguration(config), "Invalid sharding configuration in AlgorithmProvidedShardingRuleConfiguration.");
         return new ShardingRule(config, dataSources.keySet(), instanceContext);
+    }
+    
+    private boolean isValidRuleConfiguration(final AlgorithmProvidedShardingRuleConfiguration config) {
+        Map<String, KeyGenerateAlgorithm> keyGenerators = config.getKeyGenerators();
+        Map<String, ShardingAlgorithm> shardingAlgorithms = config.getShardingAlgorithms();
+        if (isInvalidKeyGenerateStrategy(config.getDefaultKeyGenerateStrategy(), keyGenerators)
+                || isInvalidShardingStrategy(config.getDefaultDatabaseShardingStrategy(), shardingAlgorithms)
+                || isInvalidShardingStrategy(config.getDefaultTableShardingStrategy(), shardingAlgorithms)) {
+            return false;
+        }
+        for (ShardingTableRuleConfiguration each : config.getTables()) {
+            if (isInvalidKeyGenerateStrategy(each.getKeyGenerateStrategy(), keyGenerators) || isInvalidShardingStrategy(each.getDatabaseShardingStrategy(), shardingAlgorithms)
+                    || isInvalidShardingStrategy(each.getTableShardingStrategy(), shardingAlgorithms)) {
+                return false;
+            }
+        }
+        for (ShardingAutoTableRuleConfiguration each : config.getAutoTables()) {
+            if (isInvalidKeyGenerateStrategy(each.getKeyGenerateStrategy(), keyGenerators) || isInvalidShardingStrategy(each.getShardingStrategy(), shardingAlgorithms)) {
+                return false;
+            }
+        }
+        return true;
+    }
+    
+    private boolean isInvalidKeyGenerateStrategy(final KeyGenerateStrategyConfiguration keyGenerateStrategy, final Map<String, KeyGenerateAlgorithm> keyGenerators) {
+        if (null == keyGenerateStrategy) {
+            return false;
+        }
+        return !keyGenerators.containsKey(keyGenerateStrategy.getKeyGeneratorName());
+    }
+    
+    private boolean isInvalidShardingStrategy(final ShardingStrategyConfiguration shardingStrategy, final Map<String, ShardingAlgorithm> shardingAlgorithms) {
+        if (null == shardingStrategy) {
+            return false;
+        }
+        return !shardingAlgorithms.containsKey(shardingStrategy.getShardingAlgorithmName());
     }
     
     @Override

--- a/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/builder/AlgorithmProvidedShardingRuleBuilder.java
+++ b/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/builder/AlgorithmProvidedShardingRuleBuilder.java
@@ -25,6 +25,7 @@ import org.apache.shardingsphere.sharding.algorithm.config.AlgorithmProvidedShar
 import org.apache.shardingsphere.sharding.api.config.rule.ShardingAutoTableRuleConfiguration;
 import org.apache.shardingsphere.sharding.api.config.rule.ShardingTableRuleConfiguration;
 import org.apache.shardingsphere.sharding.api.config.strategy.keygen.KeyGenerateStrategyConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.sharding.NoneShardingStrategyConfiguration;
 import org.apache.shardingsphere.sharding.api.config.strategy.sharding.ShardingStrategyConfiguration;
 import org.apache.shardingsphere.sharding.constant.ShardingOrder;
 import org.apache.shardingsphere.sharding.rule.ShardingRule;
@@ -78,7 +79,7 @@ public final class AlgorithmProvidedShardingRuleBuilder implements DatabaseRuleB
     }
     
     private boolean isInvalidShardingStrategy(final ShardingStrategyConfiguration shardingStrategy, final Map<String, ShardingAlgorithm> shardingAlgorithms) {
-        if (null == shardingStrategy) {
+        if (null == shardingStrategy || shardingStrategy instanceof NoneShardingStrategyConfiguration) {
             return false;
         }
         return !shardingAlgorithms.containsKey(shardingStrategy.getShardingAlgorithmName());

--- a/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/builder/ShardingRuleBuilder.java
+++ b/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/builder/ShardingRuleBuilder.java
@@ -18,10 +18,15 @@
 package org.apache.shardingsphere.sharding.rule.builder;
 
 import com.google.common.base.Preconditions;
+import org.apache.shardingsphere.infra.config.algorithm.ShardingSphereAlgorithmConfiguration;
 import org.apache.shardingsphere.infra.instance.InstanceContext;
 import org.apache.shardingsphere.infra.rule.ShardingSphereRule;
 import org.apache.shardingsphere.infra.rule.builder.database.DatabaseRuleBuilder;
 import org.apache.shardingsphere.sharding.api.config.ShardingRuleConfiguration;
+import org.apache.shardingsphere.sharding.api.config.rule.ShardingAutoTableRuleConfiguration;
+import org.apache.shardingsphere.sharding.api.config.rule.ShardingTableRuleConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.keygen.KeyGenerateStrategyConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.sharding.ShardingStrategyConfiguration;
 import org.apache.shardingsphere.sharding.constant.ShardingOrder;
 import org.apache.shardingsphere.sharding.rule.ShardingRule;
 
@@ -38,7 +43,44 @@ public final class ShardingRuleBuilder implements DatabaseRuleBuilder<ShardingRu
     public ShardingRule build(final ShardingRuleConfiguration config, final String databaseName,
                               final Map<String, DataSource> dataSources, final Collection<ShardingSphereRule> builtRules, final InstanceContext instanceContext) {
         Preconditions.checkArgument(null != dataSources && !dataSources.isEmpty(), "Data source names cannot be empty.");
+        Preconditions.checkArgument(isValidRuleConfiguration(config), "Invalid sharding configuration in ShardingRuleConfiguration.");
         return new ShardingRule(config, dataSources.keySet(), instanceContext);
+    }
+    
+    private boolean isValidRuleConfiguration(final ShardingRuleConfiguration config) {
+        Map<String, ShardingSphereAlgorithmConfiguration> keyGenerators = config.getKeyGenerators();
+        Map<String, ShardingSphereAlgorithmConfiguration> shardingAlgorithms = config.getShardingAlgorithms();
+        if (isInvalidKeyGenerateStrategy(config.getDefaultKeyGenerateStrategy(), keyGenerators)
+                || isInvalidShardingStrategy(config.getDefaultDatabaseShardingStrategy(), shardingAlgorithms)
+                || isInvalidShardingStrategy(config.getDefaultTableShardingStrategy(), shardingAlgorithms)) {
+            return false;
+        }
+        for (ShardingTableRuleConfiguration each : config.getTables()) {
+            if (isInvalidKeyGenerateStrategy(each.getKeyGenerateStrategy(), keyGenerators) || isInvalidShardingStrategy(each.getDatabaseShardingStrategy(), shardingAlgorithms)
+                    || isInvalidShardingStrategy(each.getTableShardingStrategy(), shardingAlgorithms)) {
+                return false;
+            }
+        }
+        for (ShardingAutoTableRuleConfiguration each : config.getAutoTables()) {
+            if (isInvalidKeyGenerateStrategy(each.getKeyGenerateStrategy(), keyGenerators) || isInvalidShardingStrategy(each.getShardingStrategy(), shardingAlgorithms)) {
+                return false;
+            }
+        }
+        return true;
+    }
+    
+    private boolean isInvalidKeyGenerateStrategy(final KeyGenerateStrategyConfiguration keyGenerateStrategy, final Map<String, ShardingSphereAlgorithmConfiguration> keyGenerators) {
+        if (null == keyGenerateStrategy) {
+            return false;
+        }
+        return !keyGenerators.containsKey(keyGenerateStrategy.getKeyGeneratorName());
+    }
+    
+    private boolean isInvalidShardingStrategy(final ShardingStrategyConfiguration shardingStrategy, final Map<String, ShardingSphereAlgorithmConfiguration> shardingAlgorithms) {
+        if (null == shardingStrategy) {
+            return false;
+        }
+        return !shardingAlgorithms.containsKey(shardingStrategy.getShardingAlgorithmName());
     }
     
     @Override

--- a/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/builder/ShardingRuleBuilder.java
+++ b/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/rule/builder/ShardingRuleBuilder.java
@@ -26,6 +26,7 @@ import org.apache.shardingsphere.sharding.api.config.ShardingRuleConfiguration;
 import org.apache.shardingsphere.sharding.api.config.rule.ShardingAutoTableRuleConfiguration;
 import org.apache.shardingsphere.sharding.api.config.rule.ShardingTableRuleConfiguration;
 import org.apache.shardingsphere.sharding.api.config.strategy.keygen.KeyGenerateStrategyConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.sharding.NoneShardingStrategyConfiguration;
 import org.apache.shardingsphere.sharding.api.config.strategy.sharding.ShardingStrategyConfiguration;
 import org.apache.shardingsphere.sharding.constant.ShardingOrder;
 import org.apache.shardingsphere.sharding.rule.ShardingRule;
@@ -77,7 +78,7 @@ public final class ShardingRuleBuilder implements DatabaseRuleBuilder<ShardingRu
     }
     
     private boolean isInvalidShardingStrategy(final ShardingStrategyConfiguration shardingStrategy, final Map<String, ShardingSphereAlgorithmConfiguration> shardingAlgorithms) {
-        if (null == shardingStrategy) {
+        if (null == shardingStrategy || shardingStrategy instanceof NoneShardingStrategyConfiguration) {
             return false;
         }
         return !shardingAlgorithms.containsKey(shardingStrategy.getShardingAlgorithmName());

--- a/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/rule/builder/AlgorithmProvidedShardingRuleBuilderTest.java
+++ b/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/rule/builder/AlgorithmProvidedShardingRuleBuilderTest.java
@@ -21,7 +21,10 @@ import org.apache.shardingsphere.infra.instance.InstanceContext;
 import org.apache.shardingsphere.infra.rule.builder.database.DatabaseRuleBuilder;
 import org.apache.shardingsphere.infra.rule.builder.database.DatabaseRuleBuilderFactory;
 import org.apache.shardingsphere.sharding.algorithm.config.AlgorithmProvidedShardingRuleConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.keygen.KeyGenerateStrategyConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.sharding.StandardShardingStrategyConfiguration;
 import org.apache.shardingsphere.sharding.rule.ShardingRule;
+import org.apache.shardingsphere.sharding.spi.KeyGenerateAlgorithm;
 import org.junit.Before;
 import org.junit.Test;
 
@@ -49,19 +52,47 @@ public final class AlgorithmProvidedShardingRuleBuilderTest {
     @SuppressWarnings("unchecked")
     @Test
     public void assertBuild() {
-        assertThat(builder.build(ruleConfig, "test_schema",
+        assertThat(builder.build(ruleConfig, "sharding_db",
                 Collections.singletonMap("name", mock(DataSource.class, RETURNS_DEEP_STUBS)), Collections.emptyList(), mock(InstanceContext.class)), instanceOf(ShardingRule.class));
     }
     
     @SuppressWarnings("unchecked")
     @Test(expected = IllegalArgumentException.class)
     public void assertBuildWithNullDataSourceMap() {
-        assertThat(builder.build(ruleConfig, "test_schema", null, Collections.emptyList(), mock(InstanceContext.class)), instanceOf(ShardingRule.class));
+        assertThat(builder.build(ruleConfig, "sharding_db", null, Collections.emptyList(), mock(InstanceContext.class)), instanceOf(ShardingRule.class));
     }
     
     @SuppressWarnings("unchecked")
     @Test(expected = IllegalArgumentException.class)
     public void assertBuildWithEmptyDataSourceMap() {
-        assertThat(builder.build(ruleConfig, "test_schema", Collections.emptyMap(), Collections.emptyList(), mock(InstanceContext.class)), instanceOf(ShardingRule.class));
+        assertThat(builder.build(ruleConfig, "sharding_db", Collections.emptyMap(), Collections.emptyList(), mock(InstanceContext.class)), instanceOf(ShardingRule.class));
+    }
+    
+    @SuppressWarnings("unchecked")
+    @Test(expected = IllegalArgumentException.class)
+    public void assertBuildWithInvalidKeyGenerator() {
+        ruleConfig.setDefaultKeyGenerateStrategy(new KeyGenerateStrategyConfiguration("order_id", "snowflake"));
+        assertThat(builder.build(ruleConfig, "sharding_db", Collections.singletonMap("ds_0", mock(DataSource.class)), Collections.emptyList(), mock(InstanceContext.class)),
+                instanceOf(ShardingRule.class));
+    }
+    
+    @SuppressWarnings("unchecked")
+    @Test(expected = IllegalArgumentException.class)
+    public void assertBuildWithInvalidDatabaseShardingStrategy() {
+        ruleConfig.setDefaultKeyGenerateStrategy(new KeyGenerateStrategyConfiguration("order_id", "snowflake"));
+        ruleConfig.getKeyGenerators().put("snowflake", mock(KeyGenerateAlgorithm.class));
+        ruleConfig.setDefaultDatabaseShardingStrategy(new StandardShardingStrategyConfiguration("user_id", "database_inline"));
+        assertThat(builder.build(ruleConfig, "sharding_db", Collections.singletonMap("ds_0", mock(DataSource.class)), Collections.emptyList(), mock(InstanceContext.class)),
+                instanceOf(ShardingRule.class));
+    }
+    
+    @SuppressWarnings("unchecked")
+    @Test(expected = IllegalArgumentException.class)
+    public void assertBuildWithInvalidTableShardingStrategy() {
+        ruleConfig.setDefaultKeyGenerateStrategy(new KeyGenerateStrategyConfiguration("order_id", "snowflake"));
+        ruleConfig.getKeyGenerators().put("snowflake", mock(KeyGenerateAlgorithm.class));
+        ruleConfig.setDefaultTableShardingStrategy(new StandardShardingStrategyConfiguration("order_id", "t_order_inline"));
+        assertThat(builder.build(ruleConfig, "sharding_db", Collections.singletonMap("ds_0", mock(DataSource.class)), Collections.emptyList(), mock(InstanceContext.class)),
+                instanceOf(ShardingRule.class));
     }
 }

--- a/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/rule/builder/ShardingRuleBuilderTest.java
+++ b/shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/test/java/org/apache/shardingsphere/sharding/rule/builder/ShardingRuleBuilderTest.java
@@ -17,10 +17,13 @@
 
 package org.apache.shardingsphere.sharding.rule.builder;
 
+import org.apache.shardingsphere.infra.config.algorithm.ShardingSphereAlgorithmConfiguration;
 import org.apache.shardingsphere.infra.instance.InstanceContext;
 import org.apache.shardingsphere.infra.rule.builder.database.DatabaseRuleBuilder;
 import org.apache.shardingsphere.infra.rule.builder.database.DatabaseRuleBuilderFactory;
 import org.apache.shardingsphere.sharding.api.config.ShardingRuleConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.keygen.KeyGenerateStrategyConfiguration;
+import org.apache.shardingsphere.sharding.api.config.strategy.sharding.StandardShardingStrategyConfiguration;
 import org.apache.shardingsphere.sharding.rule.ShardingRule;
 import org.junit.Before;
 import org.junit.Test;
@@ -49,19 +52,47 @@ public final class ShardingRuleBuilderTest {
     @SuppressWarnings("unchecked")
     @Test
     public void assertBuild() {
-        assertThat(builder.build(ruleConfig, "test_schema",
+        assertThat(builder.build(ruleConfig, "sharding_db",
                 Collections.singletonMap("name", mock(DataSource.class, RETURNS_DEEP_STUBS)), Collections.emptyList(), mock(InstanceContext.class)), instanceOf(ShardingRule.class));
     }
     
     @SuppressWarnings("unchecked")
     @Test(expected = IllegalArgumentException.class)
     public void assertBuildWithNullDataSourceMap() {
-        assertThat(builder.build(ruleConfig, "test_schema", null, Collections.emptyList(), mock(InstanceContext.class)), instanceOf(ShardingRule.class));
+        assertThat(builder.build(ruleConfig, "sharding_db", null, Collections.emptyList(), mock(InstanceContext.class)), instanceOf(ShardingRule.class));
     }
     
     @SuppressWarnings("unchecked")
     @Test(expected = IllegalArgumentException.class)
     public void assertBuildWithEmptyDataSourceMap() {
-        assertThat(builder.build(ruleConfig, "test_schema", Collections.emptyMap(), Collections.emptyList(), mock(InstanceContext.class)), instanceOf(ShardingRule.class));
+        assertThat(builder.build(ruleConfig, "sharding_db", Collections.emptyMap(), Collections.emptyList(), mock(InstanceContext.class)), instanceOf(ShardingRule.class));
+    }
+    
+    @SuppressWarnings("unchecked")
+    @Test(expected = IllegalArgumentException.class)
+    public void assertBuildWithInvalidKeyGenerator() {
+        ruleConfig.setDefaultKeyGenerateStrategy(new KeyGenerateStrategyConfiguration("order_id", "snowflake"));
+        assertThat(builder.build(ruleConfig, "sharding_db", Collections.singletonMap("ds_0", mock(DataSource.class)), Collections.emptyList(), mock(InstanceContext.class)),
+                instanceOf(ShardingRule.class));
+    }
+    
+    @SuppressWarnings("unchecked")
+    @Test(expected = IllegalArgumentException.class)
+    public void assertBuildWithInvalidDatabaseShardingStrategy() {
+        ruleConfig.setDefaultKeyGenerateStrategy(new KeyGenerateStrategyConfiguration("order_id", "snowflake"));
+        ruleConfig.getKeyGenerators().put("snowflake", mock(ShardingSphereAlgorithmConfiguration.class));
+        ruleConfig.setDefaultDatabaseShardingStrategy(new StandardShardingStrategyConfiguration("user_id", "database_inline"));
+        assertThat(builder.build(ruleConfig, "sharding_db", Collections.singletonMap("ds_0", mock(DataSource.class)), Collections.emptyList(), mock(InstanceContext.class)),
+                instanceOf(ShardingRule.class));
+    }
+    
+    @SuppressWarnings("unchecked")
+    @Test(expected = IllegalArgumentException.class)
+    public void assertBuildWithInvalidTableShardingStrategy() {
+        ruleConfig.setDefaultKeyGenerateStrategy(new KeyGenerateStrategyConfiguration("order_id", "snowflake"));
+        ruleConfig.getKeyGenerators().put("snowflake", mock(ShardingSphereAlgorithmConfiguration.class));
+        ruleConfig.setDefaultTableShardingStrategy(new StandardShardingStrategyConfiguration("order_id", "t_order_inline"));
+        assertThat(builder.build(ruleConfig, "sharding_db", Collections.singletonMap("ds_0", mock(DataSource.class)), Collections.emptyList(), mock(InstanceContext.class)),
+                instanceOf(ShardingRule.class));
     }
 }

--- a/shardingsphere-test/shardingsphere-integration-driver-test/src/test/resources/yaml/integrate/sharding/configWithoutDataSourceWithProps.yaml
+++ b/shardingsphere-test/shardingsphere-integration-driver-test/src/test/resources/yaml/integrate/sharding/configWithoutDataSourceWithProps.yaml
@@ -61,9 +61,7 @@ rules:
   defaultDatabaseStrategy:
     none:
   defaultTableStrategy:
-    complex:
-      shardingColumns: id, order_id
-      shardingAlgorithmName: complex_sharding_fixture
+    none:
   
   shardingAlgorithms:
     standard_sharding_fixture:

--- a/shardingsphere-test/shardingsphere-pipeline-test/src/test/resources/config_sharding_sphere_jdbc_source.yaml
+++ b/shardingsphere-test/shardingsphere-pipeline-test/src/test/resources/config_sharding_sphere_jdbc_source.yaml
@@ -49,6 +49,9 @@ rules:
       type: INLINE
       props:
         algorithm-expression: t_order
+  keyGenerators:
+    snowflake:
+      type: SNOWFLAKE
   scalingName: default_scaling
   scaling:
     default_scaling:

